design-documents/graph-ql/nullability.md

```diff
@@ -0,0 +1,206 @@
+
+_[Further reading on the relationship between non-null types and errors](http://spec.graphql.org/draft/#sec-Errors-and-Non-Nullability)_
+
+### Example: Bad use of non-nullable field
+
+In this example, a client wants to query for data to render a product details page. But, because the `related_products` field is non-nullable, the client will lose access to data like `name` and `price` if `related_products` throws any errors.
+
+Ideally, a UI would get the critical product data, render the page, and just not render the "Related Products" display.
+
+**Schema**:
+```graphql
+type Query {
+   product(id: ID): Product
+}
+
+type Product {
+	id: ID!
+	name: String!
+    price: Money!
+    description: String!
+    url: String!
+    related_products: RelatedProducts!
+}
+```
+**Query**:
+```graphql
+query ProductDetailsPage($id: ID) {
+	product(id: $id) {
+       id
+       name
+       price
+       description
+       # related_products can't be null. When the backing service is
+       # down/unreachable at the time this query runs, we can't fulfill
+       # the contract
+       related_products {
+           items {
+               id
+               name
+               price
+               url
+           }
+       }
+    }
+}
+```
+
+**Response**
+
+Note that no data was obtained by the client for `product`
+```js
+{
+   "data": {
+      "product": null
+   },
+   "errors": [
+      // details about failure in related_products field populated here
+	]
+}
+```
+
+## Backwards Compatibility
+
+- It is a safe, non-breaking change to move a field from nullable to non-nullable
+- It is an unsafe, breaking change to move a field from non-nullable to nullable
+
+With this in mind, when in doubt, the safer route is to make a field nullable, with the agreement that we'll iterate and improve schemas as we get feedback on usage from clients.
+
+**Note**: The [GraphQL reference implementation](https://github.com/graphql/graphql-js) has a `findBreakingChanges` feature that can be inspected to see a [list of all known schema changes that cause breaks for clients](https://github.com/graphql/graphql-js/blob/2232ebdef828566f3add3ed2a31709d3c1710c0e/src/utilities/findBreakingChanges.js#L41-L67). It [explicitly notes](https://github.com/graphql/graphql-js/blob/2232ebdef828566f3add3ed2a31709d3c1710c0e/src/utilities/findBreakingChanges.js#L461) that moving from a nullable value to a non-nullable value of the same type is safe.
+
+## Recommendations
+
+### Top-level Query fields should always be nullable
+
+Because of error propagation, if an error occurs in a non-null field on the `Query` root, the client will lose data from all other top-level queries.
+
+```graphql
+type Query {
+    # If a client queries for both fields, and one of them fails,
+    # the client will receive no data
+    product(id: ID): Product!
+    category(id: ID): Category!
+}
+```
```

Comment:
Hi @DrewML much needed document! Can you also mention best practices for subqueries. Can we generalize that the return type of any query, be it a parent or a subquery, can be nullable? 
I see some conflicts in the upcoming features
<img width="318" alt="Screen Shot 2020-07-06 at 4 56 30 PM" src="https://user-images.githubusercontent.com/43612864/86652382-fb9d7500-bfa9-11ea-86ad-00610b142676.png">

Reply:
Thanks @prabhuram93 for the review and feedback!

My hope/intent was that the case you've raised would be covered by this section:

![image](https://user-images.githubusercontent.com/5233399/86656411-11f90000-bfad-11ea-9a36-4d639ae8ae2f.png)

In your example screenshot, I can think of many cases where a response for part of `Customer` would still be valuable if `Customer.reviews` failed, so ideally `Customer.reviews` _would_ be nullable.

Can you think of what wording we might be able to change to make that stand out more?

Reply:
The recommendations are clear and does stand out. Since you were very specific about Top level Queries, I just wanted to imply/consult that the same is applicable for queries at any level.

Reply:
> I just wanted to imply/consult that the same is applicable for queries at any level

Definitely! 😄 

I tried to avoid getting too prescriptive about nested fields because the rules there depend a lot more on the domain object (i.e. is a product that has options useful without its options data?).

The reason for going prescriptive with root queries was because errors bubbling up to that point and finding a non-null would lose other, likely unrelated data.

More than happy to add in some clarifications if you think they'd help though! Hard not to have tunnel vision with your own writing 😅